net-libs/grpc/grpc-1.12.1.ebuild

```diff
@@ -0,0 +1,154 @@
+# Copyright 1999-2018 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=6
+
+PYTHON_COMPAT=( python2_7 python3_{4,5,6} )
+DISTUTILS_OPTIONAL=1
+
+inherit distutils-r1 flag-o-matic toolchain-funcs
+
+PROTOBUF_VERSION="3.5.2"
+
+DESCRIPTION="Modern open source high performance RPC framework"
+HOMEPAGE="http://www.grpc.io"
+SRC_URI="
+	https://github.com/${PN}/${PN}/archive/v${PV}.tar.gz -> ${P}.tar.gz
+	tools? ( https://github.com/google/protobuf/archive/v${PROTOBUF_VERSION}.tar.gz -> protobuf-${PROTOBUF_VERSION}.tar.gz )
+"
+
+LICENSE="Apache-2.0"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="examples doc python systemtap tools"
+
+REQUIRED_USE="
+	python? ( ${PYTHON_REQUIRED_USE} )
+	tools? ( python )
+"
+
+DEPEND="
+	>=dev-libs/openssl-1.0.2:0=[-bindist]
+	>=dev-libs/protobuf-3.5:=
+	dev-util/google-perftools
+	net-dns/c-ares:=
+	sys-libs/zlib:=
+	python? ( ${PYTHON_DEPS}
+		dev-python/coverage[${PYTHON_USEDEP}]
+		dev-python/cython[${PYTHON_USEDEP}]
+		>=dev-python/protobuf-python-3.5.1:=[${PYTHON_USEDEP}]
+		dev-python/six[${PYTHON_USEDEP}]
+		dev-python/wheel[${PYTHON_USEDEP}]
+		virtual/python-enum34[${PYTHON_USEDEP}]
+		virtual/python-futures[${PYTHON_USEDEP}]
+		doc? (
+			dev-python/sphinx[${PYTHON_USEDEP}]
+			dev-python/sphinx_rtd_theme[${PYTHON_USEDEP}]
+		)
+	)
+	systemtap? ( dev-util/systemtap )
+"
+
+RDEPEND="${DEPEND}"
+
+PATCHES=(
+	"${FILESDIR}/0001-grpc-1.11.0-Fix-cross-compiling.patch"
+	"${FILESDIR}/0002-grpc-1.3.0-Fix-unsecure-.pc-files.patch"
+	"${FILESDIR}/0003-grpc-1.3.0-Don-t-run-ldconfig.patch"
+	"${FILESDIR}/0004-grpc-1.11.0-fix-cpp-so-version.patch"
+	"${FILESDIR}/0005-grpc-1.11.0-pkgconfig-libdir.patch"
+	"${FILESDIR}/0006-grpc-1.12.1-allow-system-openssl.patch"
+	"${FILESDIR}/0007-grpc-1.12.1-allow-system-zlib.patch"
+	"${FILESDIR}/0008-grpc-1.12.1-allow-system-cares.patch"
+)
+
+src_prepare() {
+	sed -i 's@$(prefix)/lib@$(prefix)/$(INSTALL_LIBDIR)@g' Makefile || die
+	default
+	use python && distutils-r1_src_prepare
+}
+
+python_prepare() {
+	if use tools; then
+		rm -r third_party/protobuf || die
+		ln -s "${S}"/../protobuf-"${PROTOBUF_VERSION}" third_party/protobuf || die
```

Comment:
they bundle protobuf in the tarball i take it? and you're nuking it and replacing with the portage-downloaded tarball?
thats good perhaps add a small comment.

but more importantly, I wonder if we should be doing the rm -r third_party/protobuf unconditionally in src_prepare and then optionally do all this symlinking before the distutils-r1_src_prepare. then this whole python_prepare function could be skipped and left as default

Reply:
it's a git submodule. Just nuking empty dir.
I fetch exactly submodule specified version, it also matches latest portage version.
and grpio-tools needs both protobuf headers and source.
I had it almost build with system-protobuf by pointing to headers, but it still need some cpp files which are available only in source.

Reply:
and just to clarify. third_party/protobuf is empty in a tarball, but is a git submodule dir on git checkout. so just removing empty dir to symlink unpacked distfile there.
I added a die comment, hopefully it's enough.